File: reference-implementations/typescript/ValidationResult.ts
/**
 * ValidationResult - Immutable DTO for validation results
 *
 * NORP Compliance:
 * - NORP-003: Immutable DTO (readonly properties)
 * - NORP-001: Validation result structure
 * - NORP-007: Cost estimation included
 *
 * @license MIT
 * @copyright 2026 NeuraScope CONVERWAY
 */

export interface ValidationResult {
  readonly valid: boolean;
  readonly errors: ReadonlyArray<string>;
  readonly warnings: ReadonlyArray<string>;
  readonly estimated_cost: number;
}

export class ValidationResultImpl implements ValidationResult {
  readonly valid: boolean;
  readonly errors: ReadonlyArray<string>;
  readonly warnings: ReadonlyArray<string>;
  readonly estimated_cost: number;

  constructor(
    valid: boolean,
    errors: string[] = [],
    warnings: string[] = [],
    estimated_cost: number = 0.0
  ) {
    this.valid = valid;
    this.errors = Object.freeze([...errors]);
    this.warnings = Object.freeze([...warnings]);
    this.estimated_cost = estimated_cost;

    // NORP-003: Deep immutability
    Object.freeze(this);
  }

  hasCriticalErrors(): boolean {
    return !this.valid && this.errors.length > 0;
  }

  toJSON(): object {
    return {
      valid: this.valid,
      errors: this.errors,
      warnings: this.warnings,
      estimated_cost: this.estimated_cost
    };
  }

  getSummary(): string {
    if (this.valid) {
      let msg = 'Validation passed';
      if (this.warnings.length > 0) {
        msg += ` (${this.warnings.length} warnings)`;
      }
      return msg;
    }

    return 'Validation failed: ' + this.errors.join(', ');
  }
}

File: reference-implementations/python/validation_result.py
"""
ValidationResult - Immutable DTO for validation results

NORP Compliance:
- NORP-003: Immutable DTO (frozen dataclass)
- NORP-001: Validation result structure
- NORP-007: Cost estimation included

License: MIT
Copyright: 2026 NeuraScope CONVERWAY
"""

from dataclasses import dataclass
from typing import List


@dataclass(frozen=True)
class ValidationResult:
    """
    Immutable validation result (NORP-003 compliant)
    """
    valid: bool
    errors: List[str]
    warnings: List[str]
    estimated_cost: float = 0.0

    def has_critical_errors(self) -> bool:
        """Check if critical errors are present"""
        return not self.valid and len(self.errors) > 0

    def to_dict(self) -> dict:
        """Convert to dictionary for serialization"""
        return {
            'valid': self.valid,
            'errors': self.errors,
            'warnings': self.warnings,
            'estimated_cost': self.estimated_cost
        }

    def get_summary(self) -> str:
        """Get summary message"""
        if self.valid:
            msg = 'Validation passed'
            if self.warnings:
                msg += f' ({len(self.warnings)} warnings)'
            return msg

        return 'Validation failed: ' + ', '.join(self.errors)
